datafusion-jit/src/api.rs

```diff
@@ -0,0 +1,630 @@
+    pub fn register_extern_fn(
+        &self,
+        name: impl Into<String>,
+        ptr: *const u8,
+        params: Vec<JITType>,
+        returns: Option<JITType>,
+    ) -> Result<()> {
+        let extern_funcs = &mut self.state.lock().extern_funcs;
+        let fn_name = name.into();
+        let old = extern_funcs.insert(
+            fn_name.clone(),
+            ExternFuncSignature {
+                name: fn_name,
+                code: ptr,
+                params,
+                returns,
+            },
+        );
+
+        match old {
+            None => Ok(()),
+            Some(old) => internal_err!("Extern function {} already exists", old.name),
+        }
+    }
+
+    /// Create a new FunctionBuilder with `name` prefix
+    pub fn new_func_builder(&self, name: impl Into<String>) -> FunctionBuilder {
+        let name = self.state.lock().fresh_name(name);
+        FunctionBuilder::new(name, self.state.clone())
+    }
+
+    /// Create JIT env which we could compile the AST of constructed function
+    /// into runnable code.
+    pub fn create_jit(&self) -> JIT {
+        let symbols = self
+            .state
+            .lock()
+            .extern_funcs
+            .values()
+            .map(|s| (s.name.clone(), s.code))
+            .collect::<Vec<_>>();
+        JIT::new(symbols)
+    }
+}
+
+/// Function builder API. Stores the state while
+/// we are constructing an AST for a function.
+pub struct FunctionBuilder {
+    name: String,
+    params: Vec<(String, JITType)>,
+    ret: Option<(String, JITType)>,
+    fields: VecDeque<HashMap<String, JITType>>,
+    assembler_state: Arc<Mutex<AssemblerState>>,
+}
+
+impl FunctionBuilder {
+    fn new(name: impl Into<String>, assembler_state: Arc<Mutex<AssemblerState>>) -> Self {
+        let mut fields = VecDeque::new();
+        fields.push_back(HashMap::new());
+        Self {
+            name: name.into(),
+            params: Vec::new(),
+            ret: None,
+            fields,
+            assembler_state,
+        }
+    }
+
+    /// Add one more parameter to the function.
+    pub fn param(mut self, name: impl Into<String>, ty: JITType) -> Self {
+        let name = name.into();
+        assert!(!self.fields.back().unwrap().contains_key(&name));
+        self.params.push((name.clone(), ty));
+        self.fields.back_mut().unwrap().insert(name, ty);
+        self
+    }
+
+    /// Set return type for the function. Functions are of `void` type by default if
+    /// you do not set the return type.
+    pub fn ret(mut self, name: impl Into<String>, ty: JITType) -> Self {
+        let name = name.into();
+        assert!(!self.fields.back().unwrap().contains_key(&name));
+        self.ret = Some((name.clone(), ty));
+        self.fields.back_mut().unwrap().insert(name, ty);
+        self
+    }
+
+    /// Enter the function body at start the building.
+    pub fn enter_block(&mut self) -> CodeBlock {
```

Comment:
For one `FunctionBuilder`, `enter_block` can only be called once at the beginning of building, right?

Reply:
Yes, right after declaring function parameter and returns.